src/games/strategy/triplea/delegate/MoveValidator.java

```diff
@@ -1430,8 +1430,10 @@ public static String validateCanal(final Route route, final Collection<Unit> uni
         continue;
       }
       failureMessage = canPassThroughCanal(canalAttachment, units, player, data);
-      if (!failureMessage.isPresent()) {
```

Comment:
The method took a *lot* of work to understand. It can be simplified to;
```
    return CanalAttachment.get(territory).stream()
                                  .filter( canalAttachment ->  isCanalOnRoute(canalAttachment, route, data) )
                                  .filter( canalAttachment ->  canPassThroughCanal(canalAttachment, units, player, data).isPresent())
                                  .findFirst();
```

One big problem is that `canPassThroughCanal` is misnamed. Renaming that, yields something that makes sense more readily:
    ```
    return CanalAttachment.get(territory).stream()
        .filter( canalAttachment ->  isCanalOnRoute(canalAttachment, route, data) )
        .filter( canalAttachment ->  findErrorsWhenPassingThroughCanal(canalAttachment, units, player, data).isPresent())
        .findFirst();
    ```